src/lib/coreaudio/CoreAudioDevice.cxx

```diff
@@ -0,0 +1,676 @@
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "CoreAudioDevice.hxx"
+#include "CoreAudioHelpers.hxx"
+#include "CoreAudioHardware.hxx"
+#include "util/StringBuffer.hxx"
+#include <AudioToolbox/AudioToolbox.h>
+#include "system/ByteOrder.hxx"
+#include "util/Domain.hxx"
+#include "Log.hxx"
+
+static constexpr Domain macos_output_domain("macos_output");
+
+CoreAudioDevice::CoreAudioDevice()  :
+	started					(false	),
+	device_id				(0		),
+	io_proc					(nullptr),
+	has_volume				(false	),
+	hog_pid					(-1		),
+	buffer_size_restore		(0		)
+{
+	output_format.mFormatID = 0;
+}
+
+CoreAudioDevice::CoreAudioDevice(AudioDeviceID dev_id) :
+	started					(false		),
+	device_id				(dev_id		),
+	io_proc					(nullptr	),
+	has_volume				(false		),
+	hog_pid					(-1			),
+	buffer_size_restore		(0			)
+{
+	output_format.mFormatID = 0;
+}
+
+CoreAudioDevice::~CoreAudioDevice()
+{
+	Close();
+}
+
+bool
+CoreAudioDevice::Open(const std::string &device_name)
+{
+	if(!(device_id = CCoreAudioHardware::FindAudioDevice(device_name)))
+		return false;
+	
+	if(!Enumerate())
+	{
+		FormatError(macos_output_domain, "Failed to get stream information from output device.");
+		return false;
+	}
+
+	AudioObjectPropertyAddress aopa;
+	aopa.mScope		= kAudioDevicePropertyScopeOutput;
+	aopa.mElement	= 0;
+	aopa.mSelector	= kAudioHardwareServiceDeviceProperty_VirtualMasterVolume;
+	Boolean writable;
+	
+	if(AudioObjectHasProperty(device_id, &aopa))
+	{
+		OSStatus err = AudioObjectIsPropertySettable(device_id, &aopa, &writable);
+		if (err != noErr)
+		{
+			FormatError(macos_output_domain, "Unable to get propertyinfo volume support. Error = %s", GetError(err).c_str());
+			has_volume = false;
+		}
+		else
+			has_volume = (bool)writable;
+	}
+	else
+	{
+		has_volume = false;
+		FormatInfo(macos_output_domain, "The audio device (id 0x%04x) does not have Volume property.", (uint)device_id);
+	}
+	return true;
+}
+
+bool
+CoreAudioDevice::Enumerate()
+{
+	AudioStreamIdList stream_list;
+	bool ret = false;
+	stream_infos.clear();
+	is_planar = true;
+	
+	if (GetStreams(&stream_list))
+	{
+		for (UInt32 stream_idx = 0; stream_idx < stream_list.size(); stream_idx++)
+		{
+			stream_info info;
+			info.stream_id = stream_list[stream_idx];
+			info.num_channels = GetNumChannelsOfStream(stream_idx);
+			// one stream with num channels other than 1 is enough to make this device non-planar
+			if (info.num_channels != 1)
+				is_planar = false;
+			CoreAudioStream::GetAvailablePhysicalFormats(stream_list[stream_idx], &info.format_list);
+			stream_infos.push_back(info);
+		}
+		ret = true;
+	}
+	return ret;
+}
+
+void
+CoreAudioDevice::Close()
+{
+	if (!device_id)
+		return;
+
+	// Stop the device if it was started
+	Stop();
+
+	// Unregister the IOProc if available
+	RemoveIOProc();
+
+	// Make sure to release hog state
+	SetHogStatus(false);
```

Comment:
What if this function throws? Will this leak memory? (If `SetHotStatus()` cannot throw, it should be declared that way.)
This "problem" probably affects much of this patch.